File: rankWithTies/test_rankWithTies.m
% test_rankWithTies - Test suite
%{ 
%-------------------------------------------------------------------------------
% SYNTAX:
%   This code provides the test suite.  It can be run through the MOxUnit
%   unit testing framework.
%
% PURPOSE:
%   This function provides the unit tests for cconv.m  New bug reports for the
%   code should not be closed with ensuring that a test in the suite both fails
%   before the fix, and passes after it.
%
%-------------------------------------------------------------------------------
%}
function test_rankWithTies()

try % assignment of 'localfunctions' is necessary in Matlab >= 2016
    test_functions = localfunctions();
catch % no problem; early Matlab versions can use initTestSuite fine
end
initTestSuite;

end



% Test that all of the described syntax is supported
function test_syntaxSupport()

ACell = {{[pi 1 1 2 3 3 ]}};
methodCell = {{'min'}, {'max'}, {'random'}, {'stable'}, {@(rMax,nMatch)rMax} };
sortOptsCell = {{'descend'},{}};

 % Iterate ensuring every permutation of calling syntaxes execute without error
 argSize = [numel(ACell) numel(methodCell) numel(sortOptsCell)];
 
 for i = 1:prod(argSize)
    argInd = ind2subVec(argSize, i, true);
    args = [...
        ACell{argInd(1)},       ...
        methodCell{argInd(2)},  ...
        sortOptsCell{argInd(3)} ];
    try
        % This shouldn't throw
        [R,B,I] = rankWithTies( args{:} );
        
    catch err
        % Tell users which case failed
        fprintf('Unsupported argument set: %d\n', i);
        
        % Give the error
        rethrow(err);
    end
 end


end


% Test type support
function test_typeSupport()

A = 'fooBar...';
ACell = {{double(A)}, {single(A)}, {uint8(A)}, {char(A)}, ...
    {logical([1 1 0 1 0 0])}};
methodCell = {{'min'}, {'max'}, {'random'}, {'stable'}, {@(rMax,nMatch)rMax} };
sortOptsCell = {{'descend'},{}};

 % Iterate ensuring every permutation of calling syntaxes execute without error
 argSize = [numel(ACell) numel(methodCell) numel(sortOptsCell)];
 
 for i = 1:prod(argSize)
    argInd = ind2subVec(argSize, i, true);
    args = [...
        ACell{argInd(1)},       ...
        methodCell{argInd(2)},  ...
        sortOptsCell{argInd(3)} ];
    
    try
        % This shouldn't throw
        [R,B,I] = rankWithTies( args{:} );
        
        % Verify the return types
        assertEqual( class(R), 'double' );
        assertEqual( class(B), class(ACell{argInd(1)}{:}) );
        assertEqual( class(I), 'double' );
        
    catch err
        % Tell users which case failed
        fprintf('Unsupported argument set: %d\n', i);
        
        % Give the error
        rethrow(err);
    end
 end

end


% Test dimension support
function test_dimensionSupport

A = 'fooBar...';
nA = numel(A);

for dimInd = 1:4
    % Setup A
    aSize = ones(1,max(2,dimInd));
    aSize(dimInd) = nA;
    A = reshape(A,aSize);
    
    % Confirm the dimensions match
    [R,B,I] = rankWithTies( A, 'min');
    assertEqual( size(R), aSize );
    assertEqual( size(B), aSize );
    assertEqual( size(I), aSize );
    
end

end


% Test permutation properties
function test_permutationProperties

A = 'fooBar...';
method = {'min','max','random','stable',@(rMax,nMatch) (rMax-nMatch):rMax};

for i = 1:numel(method)
    [R,B,I] = rankWithTies(A, method{i});
    assertEqual( B(R), A);
    assertEqual( A(I), B);
end

end


% Test method function equivalents
% Note: We need to carefully manage the random number generator seed
function test_methodFunctionalEquivalents

method = {'min','max','random','stable'};
equivalent = {...
    @(rMax,nMatch) rMax-nMatch, ...
    @(rMax,nMatch) rMax, ...
    @(rMax,nMatch) randperm(nMatch+1) + (rMax-nMatch-1), ...
    @(rMax,nMatch) (rMax-nMatch):rMax };
A = 'fooBar...';

for i = 1:numel(method)
    % Note the random number generator state
    rngState = rng();
    
    % Rank using method
    [R1,B1,I1] = rankWithTies( A, method{i});
    
    % Reset the random number generator state
    rng(rngState);
    
    % Rank using equivalent
    [R2,B2,I2] = rankWithTies( A, equivalent{i});
    
    % Compare
    assertEqual( R1, R2);
    assertEqual( B1, B2);
    assertEqual( I1, I2);
end

end


% Test sort options
% Note: We conspire to ensure vectorization is not required via dim
function test_sortOptions

% Default Values
A = 'fooBar...';
nA = numel(A);
method = 'min';
dimOpts = {{1}, {2}, {}};
modeOpts = {{'ascend'}, {'descend'}, {}};

% Get some true answers associated with each sort mode
expectedB = {sort(A,'ascend'), sort(A,'descend'), sort(A)};

% Iterate ensuring every permutation of calling syntaxes execute without error
 argSize = [numel(dimOpts) numel(modeOpts)];
 
 for i = 1:prod(argSize)
    argInd = ind2subVec(argSize, i, true);
    sortArgs = [ dimOpts{argInd(1)}, modeOpts{argInd(2)} ];
    
    % Be sure A is in the dimension of dimOpts (no vectorization)
    if cellfun(@isempty, dimOpts(argInd(1)))
        d = 1;
    else
        d = dimOpts{argInd(1)}{1};
    end
    ASize = ones(1,max(2,d));
    ASize(d) = nA;
    A = reshape(A,ASize);
    
    % Verify the call completes
    [R,B,I] = rankWithTies(A, method, sortArgs{:});
    
    % Verify permutation properties
    assertEqual( B(R), A);
    assertEqual( A(I), B);
    
    % Verify the expected sort result
    assertEqual( B(:), expectedB{argInd(2)}(:) );
    
 end

end


% Test vectorization support
function test_vectorization

% Default Values
method = 'min';


% Vectorization along dimension 1
ASize = [5 6];
A = randn(ASize);
for i = ASize(2):-1:1
    [RTrue(:,i), BTrue(:,i), ITrue(:,i)] = rankWithTies(A(:,i), method);
end

[R,B,I] = rankWithTies(A, method);      % Implicitly on dimension 1
assertEqual(R, RTrue);
assertEqual(B, BTrue);
assertEqual(I, ITrue);

[R,B,I] = rankWithTies(A, method, 1);   % Explicitly on dimension 1
assertEqual(R, RTrue);
assertEqual(B, BTrue);
assertEqual(I, ITrue);


% Vectorization along dimension 2
for i = ASize(1):-1:1
    [RTrue(i,:), BTrue(i,:), ITrue(i,:)] = rankWithTies(A(i,:), method);
end

[R,B,I] = rankWithTies(A, method, 2);   % Explicitly on dimension 2
assertEqual(R, RTrue);
assertEqual(B, BTrue);
assertEqual(I, ITrue);

% Trivial vectorization
[R,B,I] = rankWithTies(A, method, 3);
assertEqual(R, ones(ASize));
assertEqual(B, A);
assertEqual(I, ones(ASize));

end


% Test edge cases
function test_edgeCases

method = 'max';
A = [];
ASize = size(A);

[R,B,I] = rankWithTies(A, method);
assertEqual( size(R), ASize);
assertEqual( size(B), ASize);
assertEqual( size(I), ASize);

[R,B,I] = rankWithTies(A, method, 1);
assertEqual( size(R), ASize);
assertEqual( size(B), ASize);
assertEqual( size(I), ASize);

[R,B,I] = rankWithTies(A, method, 2);
assertEqual( size(R), ASize);
assertEqual( size(B), ASize);
assertEqual( size(I), ASize);

[R,B,I] = rankWithTies(A, method, 3);
assertEqual( size(R), ASize);
assertEqual( size(B), ASize);
assertEqual( size(I), ASize);

A = 1;
ASize = size(A);

[R,B,I] = rankWithTies(A, method);
assertEqual(R, 1);
assertEqual(B, 1);
assertEqual(I, 1);

[R,B,I] = rankWithTies(A, method, 1);
assertEqual(R, 1);
assertEqual(B, 1);
assertEqual(I, 1);

[R,B,I] = rankWithTies(A, method, 2);
assertEqual(R, 1);
assertEqual(B, 1);
assertEqual(I, 1);

[R,B,I] = rankWithTies(A, method, 3);
assertEqual(R, 1);
assertEqual(B, 1);
assertEqual(I, 1);


end

File: rankWithTies/test_rankWithTies.py
#! python3
import unittest
import numpy as np
from rankWithTies import rankWithTies

class TestCases(unittest.TestCase):
    '''
     test_rankWithTies - Test suite

     PURPOSE:
       This code provides the test suite for rankWithTies through unittest.
       New bug reports for the code should not be closed with ensuring that a
       test in the suite both fails before the fix, and passes after it.
    '''

    def test_syntaxSupport(self):
        """
        Test that all of the described syntax is supported
        """
        AList = [np.array([np.pi, 1, 1, 2, 3, 3])]
        methodList = ['min', 'max', 'random', 'stable', \
            lambda rMax, nMatch : np.arange((rMax-(nMatch-1)), rMax+1)]
        sortOptsList = [[0], [0, 'mergesort'], []]

        for A in AList:
            for method in methodList:
                for opts in sortOptsList:
                    args = [A, method, *opts]

                    try:
                        R, B, I = rankWithTies(*args)
                    except:
                        self.fail(f"Unsupported argument set: {args}")

    def test_typeSupport(self):
        '''
        Tests input/output type support
        '''
        A = 'fooBar...'
        #different datatype input arrays for ranking
        AList = [np.array([float(ord(char)) for char in A]),
                 np.array([np.float32(ord(char)) for char in A]),
                 np.array([np.uint8(ord(char)) for char in A]),
                 np.array(list(A))]
        methodList = ['min', 'max', 'random', 'stable', \
            lambda rMax, nMatch : np.arange((rMax-(nMatch-1)), rMax+1)]
        sortOptsList = [[0], [0, 'mergesort'], []]

        for A in AList:
            for method in methodList:
                for opts in sortOptsList:
                    args = [A, method, *opts]

                    R, B, I = rankWithTies(*args)

                    self.assertEqual(R.dtype, np.int64)
                    self.assertEqual(B.dtype, args[0].dtype)
                    self.assertEqual(I.dtype, np.int64)

    def test_dimensionSupport(self):
        '''
        Tests input dimensionality support
        '''

        A = np.array(list('fooBar...'))
        nA = len(A)

        for dimInd in range(0,4):
            #reshape A to be along different dimensions
            aSize = np.ones(max(2, dimInd+1), dtype = 'int64')
            aSize[dimInd] = nA
            A = A.reshape(aSize)

            #Confirm the dimensions match
            R, B, I = rankWithTies( A, 'min')

            np.testing.assert_array_equal(R.shape, aSize)
            np.testing.assert_array_equal(B.shape, aSize)
            np.testing.assert_array_equal(I.shape, aSize)

    def test_permutationProperties(self):
        '''
        Tests whether the returned perumation arrays correctly sort and unsort A
        '''

        A = np.array(list('fooBar...'))
        methodList = ['min', 'max', 'random', 'stable', \
            lambda rMax, nMatch : np.arange((rMax-(nMatch-1)), rMax+1)]

        for method in methodList:
            R, B, I = rankWithTies(A, method)
            np.testing.assert_array_equal(B[R], A)
            np.testing.assert_array_equal(A[I], B)

    def test_methodFunctionalEquivalents(self):
        '''
        Tests built-in methods' functional equivalents
        Must carefully keep track of the numpy random state for method=random
        '''

        methodList = ['min', 'max', 'random', 'stable']
        equivalent = [lambda rMax, nMatch : rMax - (nMatch-1),
                      lambda rMax, nMatch : rMax,
                      lambda rMax, nMatch : np.random.permutation(nMatch) \
                       + (rMax - (nMatch - 1)),
                      lambda rMax, nMatch : \
                        np.arange((rMax-(nMatch-1)), rMax+1)]

        A = np.array(list('fooBar...'))

        for i in range(len(methodList)):

            np.random.seed(25)
            R1, B1, I1 = rankWithTies(A, methodList[i])

            np.random.seed(25)
            R2, B2, I2 = rankWithTies(A, equivalent[i])

            np.testing.assert_array_equal(R1, R2)
            np.testing.assert_array_equal(B1, B2)
            np.testing.assert_array_equal(I1, I2)


    #Skipped testing optional inputs for now as the cases are very different
    #   in numpy since asc/desc is not an arg and order matters in numpy's
    #   sort function inputs


    def test_vectorization(self):
        '''
        Test input array vectorization support
        '''

        methodList = ['min', 'max', 'random', 'stable', \
            lambda rMax, nMatch : np.arange((rMax-(nMatch-1)), rMax+1)]

        #make sure the vectorization holds for every method
        for method in methodList:
            aSize = [5,6]
            A = np.random.random(aSize)
            RTrue = np.zeros_like(A, dtype = int)
            BTrue = np.zeros_like(A)
            ITrue = np.zeros_like(A, dtype = int)

            #go down in column order and manually do 1D rankings
            for i in np.arange(0, aSize[1])[::-1]:
                RTrue[:,i], BTrue[:,i], ITrue[:,i]= rankWithTies(A[:,i], method)

            R, B, I = rankWithTies(A, method, 0)
            np.testing.assert_array_equal(R, RTrue) #explicit along dim 0
            np.testing.assert_array_equal(B, BTrue)
            np.testing.assert_array_equal(I, ITrue)

            for i in np.arange(0, aSize[0])[::-1]:
                RTrue[i,:], BTrue[i,:], ITrue[i,:]= rankWithTies(A[i,:], method)

            R, B, I = rankWithTies(A, method)
            np.testing.assert_array_equal(R, RTrue) #implicitly along dim 1
            np.testing.assert_array_equal(B, BTrue)
            np.testing.assert_array_equal(I, ITrue)

            R, B, I = rankWithTies(A, method)
            np.testing.assert_array_equal(R, RTrue, 1) #explicitly along dim 1
            np.testing.assert_array_equal(B, BTrue, 1)
            np.testing.assert_array_equal(I, ITrue, 1)


if __name__ == '__main__':
    unittest.main()
